// @sap/cds-dk/lib/init/template/enterprise-messaging-shared/index.js
const cds = require('../../../cds')
const { read, exists } = cds.utils
const { readProject } = require('../../projectReader')
const { merge } = require('../../merge')
const { srv4, enterpriseMessaging } = require('../../registries/mta')

module.exports = class EnterpriseMessagingSharedTemplate extends require('../../plugin') {

  static help() {
    return 'messaging via shared SAP Enterprise Messaging'
  }

  options() {
    return {
      'cloudevents': {
        type: 'boolean',
        short: 'c',
        help: 'Use CloudEvents formatting.',
      }
    }
  }

  async canRun() {
    const { isJava } = readProject()
    if (isJava) throw `'cds add enterprise-messaging-shared' is not available for Java yet`
    return true
  }

  static hasInProduction(env) {
    return env.requires?.messaging?.kind === 'enterprise-messaging-amqp'
  }

  async run() {
    const project = readProject()
    const { configFile, isNodejs } = project
    if (isNodejs) {
      project.shortcut = !exists(configFile) || typeof (await read(configFile)).cds?.requires?.messaging !== 'object'
      project.cloudevents = cds.cli.options.cloudevents
      await merge(__dirname, 'files/package.json.hbs').into(configFile, { with: project })
    }
  }

  async combine() {
    const project = readProject()
    const { addMta, addHelm, addHelmUnifiedRuntime, srvPath } = project

    if (addMta) {
      const srv = srv4(srvPath)
      await merge(__dirname, 'files', 'mta.yaml.hbs').into('mta.yaml', {
        project,
        additions: [srv, enterpriseMessaging],
        relationships: [{
            insert: [enterpriseMessaging, 'name'],
            into: [srv, 'requires', 'name']
        }]
      })
    }

    if (addHelm || addHelmUnifiedRuntime) {
      await merge(__dirname, 'files/values.yaml').into('chart/values.yaml')
    }

    if (!exists('event-mesh.json')) await merge(__dirname, 'files/event-mesh.json.hbs').into('event-mesh.json', { project })
  }
}
