// @sap/cds-dk/lib/init/template/extensibility/index.js
const cds = require('../../../cds')
const { exists } = cds.utils
const { readProject } = require('../../projectReader')
const { merge } = require('../../merge')
const { MULTITENANCY } = require('../../constants').OPTIONS

module.exports = class ExtensibilityTemplate extends require('../../plugin') {

  static help() {
    return 'tenant-specific model extensibility'
  }

  requires() {
    return exists('pom.xml') ? [MULTITENANCY] : [] // REVISIT: Remove this dependency
  }

  static hasInProduction(env) {
    return !!env.requires?.extensibility
  }

  async run() {
    const project = readProject()
    const { configFile } = project
    await merge(__dirname, 'files/package.json.hbs').into(configFile, { project })
  }

  async combine() {
    const project = readProject()
    const { addXsuaa } = project
    if (addXsuaa) {
      await merge(__dirname, 'files', 'xs-security.json.hbs').into('xs-security.json', {
        project,
        additions: [
          { in: 'scopes', where: { name: '$XSAPPNAME.cds.ExtensionDeveloper' }},
          { in: 'role-templates', where: { name: 'ExtensionDeveloper' }}
        ]
      })
    }
  }
}
